### PR TITLE
Hash x

### LINKED ISSUE #846: Allow adding hash(x) as a signer of an account
This means that any transaction on the account can be authorized by knowing x.
This will allow atomic cross chain trading: https://en.bitcoin.it/wiki/Atomic_cross-chain_trading

### PR DESCRIPTION
**Pull request by vogel**
Fixes #846

### CODE REVIEW

File: src/transactions/SignatureChecker.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// Copyright 2016 Stellar Development Foundation and contributors. Licensed
```
Comment (jedmccaleb) [inline src/transactions/SignatureChecker.cpp]: Let's talk about this.

File: src/transactions/SignatureChecker.cpp [hunk 2]
```diff
@@ -43,6 +44,28 @@ bool SignatureChecker::checkSignature(AccountID const &accountID, std::vector<Si
         }
     }
 
+    auto &hashXKeyWeights = signers[SIGNER_KEY_TYPE_HASH_X];
+    for (size_t i = 0; i < mSignatures.size(); i++)
+    {
+        auto x = std::string{mSignatures[i].signature.begin(), mSignatures[i].signature.end()};
```
Comment (jedmccaleb) [inline src/transactions/SignatureChecker.cpp]: Is the preimage a string?

File: src/crypto/SecretKey.cpp [hunk 3]
```diff
@@ -327,16 +328,13 @@ PubKeyUtils::verifySig(PublicKey const& key, Signature const& signature,
 SignatureHint
 PubKeyUtils::getHint(PublicKey const& pk)
```
Comment (jedmccaleb) [inline src/crypto/SecretKey.cpp]: Just have people call SignatureUtils::getHint  directly if you want to move this function there.

File: src/crypto/SecretKey.cpp [hunk 4]
```diff
@@ -327,16 +328,13 @@ PubKeyUtils::verifySig(PublicKey const& key, Signature const& signature,
...
 {
-    SignatureHint res;
-    memcpy(res.data(), &pk.ed25519().back() - res.size() + 1, res.size());
-    return res;
+    return SignatureUtils::getHint(pk.ed25519());
 }
 
 bool
 PubKeyUtils::hasHint(PublicKey const& pk, SignatureHint const& hint)
```
Comment (jedmccaleb) [inline src/crypto/SecretKey.cpp]: This seems misnamed. Seems like it should be called `doesHintMatch`  or something

File: src/transactions/SignatureChecker.cpp (hunk 2 above)
Comment (vogel) [inline src/transactions/SignatureChecker.cpp]: using just mSignatures[i].signature now

File: src/crypto/SecretKey.cpp (hunk 3 above)
Comment (vogel) [inline src/crypto/SecretKey.cpp]: fixed

File: src/crypto/SecretKey.cpp (hunk 4 above)
Comment (vogel) [inline src/crypto/SecretKey.cpp]: renamed

**Pull request closed without merging**